invalid memory address or nil pointer dereference in agent

/kind bug

**What steps did you take and what happened:**

We deployed the following triton inferenceservice:

```yaml
apiVersion: "serving.kubeflow.org/v1beta1"
kind: "InferenceService"
metadata:
  name: "triton-mms"
spec:
  predictor:
    triton:
      args:
      - --log-verbose=1
      resources:
        limits:
          cpu: "1"
          memory: "8Gi"
        requests:
          cpu: "1"
          memory: 8Gi
```

as well as the following TrainedModel

```yaml
apiVersion: "serving.kubeflow.org/v1alpha1"
kind: "TrainedModel"
metadata:
  name: "ourmodelname"
spec:
  inferenceService: triton-mms
  model:
    framework: pytorch
    storageUri: s3:/path-to-container/ourmodelname
    memory: "1Gi"
```

within this s3 container we have stored the following mode:

```
- ourmodelname/
  |_ config.pbtxt
  |_ /1
     |_ model.pt
```

we then observe the following error in the `agent` when loading the mode:
```
{"level":"info","ts":"2021-01-26T08:45:42.264Z","caller":"agent/puller.go:121","msg":"Worker is started for layout"}
{"level":"info","ts":"2021-01-26T08:45:42.264Z","caller":"agent/puller.go:129","msg":"Downloading model from s3:/path-to-container/ourmodelname"}
{"level":"info","ts":"2021-01-26T08:45:42.264Z","caller":"agent/downloader.go:47","msg":"Downloading s3:/path-to-container/ourmodelname to model dir /mnt/models"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x169f489]

goroutine 70 [running]:
github.com/kubeflow/kfserving/pkg/agent/storage.FileExists(0xc0001a4480, 0x7b, 0x1bfab30)
	/go/src/github.com/kubeflow/kfserving/pkg/agent/storage/utils.go:27 +0x99
github.com/kubeflow/kfserving/pkg/agent.(*Downloader).DownloadModel(0xc0003eaa98, 0xc0007c5288, 0x6, 0xc0001a90e0, 0x0, 0x0)
	/go/src/github.com/kubeflow/kfserving/pkg/agent/downloader.go:49 +0x40f
github.com/kubeflow/kfserving/pkg/agent.(*Puller).modelProcessor(0xc0003eaa80, 0xc0007c5288, 0x6, 0xc0001a95c0)
	/go/src/github.com/kubeflow/kfserving/pkg/agent/puller.go:130 +0x273
created by github.com/kubeflow/kfserving/pkg/agent.(*Puller).enqueueModelOp
	/go/src/github.com/kubeflow/kfserving/pkg/agent/puller.go:89 +0x120
```

Interestingly, when we try the same with the example model `cifar10` we do not observe this issue.


**What did you expect to happen:**
Model is loaded successfully 

**Environment:**

- Knative Version: 0.17
- KFServing Version: 0.5.0-rc2

--- a/pkg/agent/downloader.go
+++ b/pkg/agent/downloader.go
@@ -23,6 +23,7 @@ import (
 	"github.com/kubeflow/kfserving/pkg/apis/serving/v1alpha1"
 	"github.com/pkg/errors"
 	"go.uber.org/zap"
+	"os"
 	"path/filepath"
 	"regexp"
 	"strings"
@@ -38,16 +39,14 @@ var SupportedProtocols = []storage.Protocol{storage.S3, storage.GCS}
 
 func (d *Downloader) DownloadModel(modelName string, modelSpec *v1alpha1.ModelSpec) error {
 	if modelSpec != nil {
-		modelUri := modelSpec.StorageURI
-		hashModelUri := hash(modelUri)
-		hashFramework := hash(modelSpec.Framework)
-		hashMemory := hash(modelSpec.Memory.String())
+		sha256 := storage.AsSha256(modelSpec)
 		successFile := filepath.Join(d.ModelDir, modelName,
-			fmt.Sprintf("SUCCESS.%s.%s.%s", hashModelUri, hashFramework, hashMemory))
-		d.Logger.Infof("Downloading %s to model dir %s", modelUri, d.ModelDir)
+			fmt.Sprintf("SUCCESS.%s", sha256))
+		d.Logger.Infof("Downloading %s to model dir %s", modelSpec.StorageURI, d.ModelDir)
 		// Download if the event there is a success file and the event is one which we wish to Download
-		if !storage.FileExists(successFile) {
-			if err := d.download(modelName, modelUri); err != nil {
+		_, err := os.Stat(successFile)
+		if os.IsNotExist(err) {
+			if err := d.download(modelName, modelSpec.StorageURI); err != nil {
 				return errors.Wrapf(err, "failed to download model")
 			}
 			file, createErr := storage.Create(successFile)
@@ -55,8 +54,10 @@ func (d *Downloader) DownloadModel(modelName string, modelSpec *v1alpha1.ModelSp
 			if createErr != nil {
 				return errors.Wrapf(createErr, "failed to create success file")
 			}
-		} else {
+		} else if err == nil {
 			d.Logger.Infof("Model successFile exists already for %s", modelName)
+		} else {
+			d.Logger.Errorf("Model successFile error %v", err)
 		}
 	}
 	return nil

--- a/pkg/agent/downloader.go
+++ b/pkg/agent/downloader.go
@@ -64,6 +64,7 @@ func (d *Downloader) DownloadModel(modelName string, modelSpec *v1alpha1.ModelSp
 			if err != nil {
 				return errors.Wrapf(createErr, "failed to write the success file")
 			}
+			d.Logger.Infof("Creating successFile %s", successFile)
 		} else if err == nil {
 			d.Logger.Infof("Model successFile exists already for %s", modelName)
 		} else {

--- a/pkg/agent/downloader.go
+++ b/pkg/agent/downloader.go
@@ -18,11 +18,13 @@ package agent
 
 import (
 	"encoding/hex"
+	"encoding/json"
 	"fmt"
 	"github.com/kubeflow/kfserving/pkg/agent/storage"
 	"github.com/kubeflow/kfserving/pkg/apis/serving/v1alpha1"
 	"github.com/pkg/errors"
 	"go.uber.org/zap"
+	"io/ioutil"
 	"os"
 	"path/filepath"
 	"regexp"
@@ -53,6 +55,14 @@ func (d *Downloader) DownloadModel(modelName string, modelSpec *v1alpha1.ModelSp
 			defer file.Close()
 			if createErr != nil {
 				return errors.Wrapf(createErr, "failed to create success file")
+			}
+			encodedJson, err := json.Marshal(modelSpec)
+			if err != nil {
+				return errors.Wrapf(createErr, "failed to encode model spec")
+			}
+			err = ioutil.WriteFile(successFile, encodedJson, 0644)
+			if err != nil {
+				return errors.Wrapf(createErr, "failed to write the success file")
 			}
 		} else if err == nil {
 			d.Logger.Infof("Model successFile exists already for %s", modelName)

--- a/pkg/agent/storage/utils.go
+++ b/pkg/agent/storage/utils.go
@@ -19,6 +19,7 @@ package storage
 import (
 	gstorage "cloud.google.com/go/storage"
 	"context"
+	"crypto/sha256"
 	"fmt"
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/session"
@@ -35,7 +36,21 @@ import (
 
 func FileExists(filename string) bool {
 	info, err := os.Stat(filename)
-	return !os.IsNotExist(err) && !info.IsDir()
+	if os.IsNotExist(err) {
+		return false
+	}
+	if err == nil && !info.IsDir() {
+		return true
+	} else {
+		return false
+	}
+}
+
+func AsSha256(o interface{}) string {
+	h := sha256.New()
+	h.Write([]byte(fmt.Sprintf("%v", o)))
+
+	return fmt.Sprintf("%x", h.Sum(nil))
 }
 
 func Create(fileName string) (*os.File, error) {

--- a/pkg/agent/syncer.go
+++ b/pkg/agent/syncer.go
@@ -17,11 +17,11 @@ limitations under the License.
 package agent
 
 import (
-	"encoding/hex"
+	"encoding/json"
 	"fmt"
 	"github.com/kubeflow/kfserving/pkg/apis/serving/v1alpha1"
 	"github.com/pkg/errors"
-	"k8s.io/apimachinery/pkg/api/resource"
+	"io/ioutil"
 	"os"
 	"path/filepath"
 	"strings"
@@ -39,17 +39,23 @@ func SyncModelDir(modelDir string) (map[string]modelWrapper, error) {
 			ierr := filepath.Walk(path, func(path string, f os.FileInfo, _ error) error {
 				if !f.IsDir() {
 					base := filepath.Base(path)
-					baseSplit := strings.SplitN(base, ".", 4)
+					baseSplit := strings.SplitN(base, ".", 2)
 					if baseSplit[0] == "SUCCESS" {
-						if spec, e := successParse(baseSplit); e != nil {
-							return errors.Wrapf(err, "error parsing SUCCESS file")
-						} else {
-							modelTracker[modelName] = modelWrapper{
-								Spec:  spec,
-								stale: true,
-							}
-							return nil
+						jsonFile, err := os.Open(path)
+						if err != nil {
+							return errors.Wrapf(err, "failed to parse success file")
 						}
+						byteValue, _ := ioutil.ReadAll(jsonFile)
+						modelSpec := &v1alpha1.ModelSpec{}
+						err = json.Unmarshal(byteValue, &modelSpec)
+						if err != nil {
+							return errors.Wrapf(err, "failed to unmarshal model spec")
+						}
+						modelTracker[modelName] = modelWrapper{
+							Spec:  modelSpec,
+							stale: true,
+						}
+						return nil
 					}
 				}
 				return NoSuccessFile
@@ -69,32 +75,3 @@ func SyncModelDir(modelDir string) (map[string]modelWrapper, error) {
 	return modelTracker, nil
 }
 
-func successParse(baseSplit []string) (*v1alpha1.ModelSpec, error) {
-	storageURI, err := unhash(baseSplit[1])
-	errorMessage := "unable to unhash the SUCCESS file, maybe the SUCCESS file has been modified?"
-	if err != nil {
-		return nil, errors.Wrapf(err, errorMessage)
-	}
-	framework, err := unhash(baseSplit[2])
-	if err != nil {
-		return nil, errors.Wrapf(err, errorMessage)
-	}
-	memory, err := unhash(baseSplit[3])
-	if err != nil {
-		return nil, errors.Wrapf(err, errorMessage)
-	}
-	memoryResource := resource.MustParse(memory)
-	return &v1alpha1.ModelSpec{
-		StorageURI: storageURI,
-		Framework:  framework,
-		Memory:     memoryResource,
-	}, nil
-}
-
-func unhash(s string) (string, error) {
-	decoded, err := hex.DecodeString(s)
-	if err != nil {
-		return "", nil
-	}
-	return string(decoded), nil
-}

--- a/pkg/agent/syncer.go
+++ b/pkg/agent/syncer.go
@@ -21,6 +21,7 @@ import (
 	"fmt"
 	"github.com/kubeflow/kfserving/pkg/apis/serving/v1alpha1"
 	"github.com/pkg/errors"
+	"go.uber.org/zap"
 	"io/ioutil"
 	"os"
 	"path/filepath"
@@ -31,40 +32,39 @@ type FileError error
 
 var NoSuccessFile FileError = fmt.Errorf("no success file can be found")
 
-func SyncModelDir(modelDir string) (map[string]modelWrapper, error) {
+func SyncModelDir(modelDir string, logger *zap.SugaredLogger) (map[string]modelWrapper, error) {
+	logger.Infof("Syncing from model dir %s", modelDir)
 	modelTracker := make(map[string]modelWrapper)
 	err := filepath.Walk(modelDir, func(path string, info os.FileInfo, err error) error {
-		if info.IsDir() {
-			modelName := info.Name()
-			ierr := filepath.Walk(path, func(path string, f os.FileInfo, _ error) error {
-				if !f.IsDir() {
-					base := filepath.Base(path)
-					baseSplit := strings.SplitN(base, ".", 2)
-					if baseSplit[0] == "SUCCESS" {
-						jsonFile, err := os.Open(path)
-						if err != nil {
-							return errors.Wrapf(err, "failed to parse success file")
-						}
-						byteValue, _ := ioutil.ReadAll(jsonFile)
-						modelSpec := &v1alpha1.ModelSpec{}
-						err = json.Unmarshal(byteValue, &modelSpec)
-						if err != nil {
-							return errors.Wrapf(err, "failed to unmarshal model spec")
-						}
-						modelTracker[modelName] = modelWrapper{
-							Spec:  modelSpec,
-							stale: true,
-						}
-						return nil
-					}
+		if !info.IsDir() {
+			fileName := info.Name()
+			if strings.HasPrefix(fileName, "SUCCESS.") {
+				logger.Infof("Syncing from model success file %v", fileName)
+				dir := filepath.Dir(path)
+				dirSplit := strings.Split(dir, "/")
+				if len(dirSplit) < 2 {
+					return errors.Wrapf(err, "invalid model path")
 				}
-				return NoSuccessFile
-			})
-			switch ierr {
-			case NoSuccessFile:
-				return nil
-			default:
-				return errors.Wrapf(ierr, "failed to parse success file")
+				modelName := dirSplit[len(dirSplit)-1]
+
+				jsonFile, err := os.Open(path)
+				if err != nil {
+					return errors.Wrapf(err, "failed to parse success file")
+				}
+				byteValue, err := ioutil.ReadAll(jsonFile)
+				if err != nil {
+					return errors.Wrapf(err, "failed to read from model spec")
+				}
+				modelSpec := &v1alpha1.ModelSpec{}
+				err = json.Unmarshal(byteValue, &modelSpec)
+				if err != nil {
+					return errors.Wrapf(err, "failed to unmarshal model spec")
+				}
+				modelTracker[dirSplit[len(dirSplit)-1]] = modelWrapper{
+					Spec:  modelSpec,
+					stale: true,
+				}
+				logger.Infof("recovered model %s with spec %+v", modelName, modelSpec)
 			}
 		}
 		return nil
@@ -74,4 +74,3 @@ func SyncModelDir(modelDir string) (map[string]modelWrapper, error) {
 	}
 	return modelTracker, nil
 }
-

--- a/pkg/agent/watcher.go
+++ b/pkg/agent/watcher.go
@@ -37,9 +37,9 @@ type Watcher struct {
 }
 
 func NewWatcher(configDir string, modelDir string, logger *zap.SugaredLogger) Watcher {
-	modelTracker, err := SyncModelDir(modelDir)
+	modelTracker, err := SyncModelDir(modelDir, logger)
 	if err != nil {
-		logger.Error(err, "Failed to sync model dir")
+		logger.Errorf("Failed to sync model dir %v", err)
 	}
 	watcher := Watcher{
 		configDir:    configDir,
@@ -50,7 +50,7 @@ func NewWatcher(configDir string, modelDir string, logger *zap.SugaredLogger) Wa
 	modelConfigFile := fmt.Sprintf("%s/%s", configDir, constants.ModelConfigFileName)
 	err = watcher.syncModelConfig(modelConfigFile)
 	if err != nil {
-		logger.Error(err, "Failed to sync model config file")
+		logger.Errorf("Failed to sync model config file %v", err)
 	}
 	return watcher
 }

--- a/pkg/agent/watcher_test.go
+++ b/pkg/agent/watcher_test.go
@@ -30,6 +30,7 @@ import (
 	. "github.com/onsi/gomega"
 	"go.uber.org/zap"
 	"io/ioutil"
+	"k8s.io/apimachinery/pkg/api/resource"
 	logger "log"
 	"os"
 	"path/filepath"
@@ -65,6 +66,7 @@ var _ = Describe("Watcher", func() {
 						Spec: v1alpha1.ModelSpec{
 							StorageURI: "s3://models/model1",
 							Framework:  "sklearn",
+							Memory:     resource.MustParse("100Mi"),
 						},
 					},
 					{
@@ -72,6 +74,7 @@ var _ = Describe("Watcher", func() {
 						Spec: v1alpha1.ModelSpec{
 							StorageURI: "s3://models/model2",
 							Framework:  "sklearn",
+							Memory:     resource.MustParse("100Mi"),
 						},
 					},
 				}
@@ -96,6 +99,8 @@ var _ = Describe("Watcher", func() {
 				Eventually(func() int { return len(puller.channelMap) }).Should(Equal(0))
 				Eventually(func() int { return puller.opStats["model1"][Add] }).Should(Equal(1))
 				Eventually(func() int { return puller.opStats["model2"][Add] }).Should(Equal(1))
+				modelSpecMap, _ := SyncModelDir(modelDir+"/test1", watcher.logger)
+				Expect(watcher.modelTracker).Should(Equal(modelSpecMap))
 			})
 		})
 	})

